### PR TITLE
Sirc(fix): Fix use-after-free when results of sirc_message_context_get_time outlive SircMessageContext

### PR DESCRIPTION
**Pull request by progval**
Closes GH-355

### CODE REVIEW

**Comment by SilverRainZ** [comment]
IMO,  the `GDateTime` would be better to be ref at `SirMessage` side?

**Review by SilverRainZ — CHANGES_REQUESTED** [review]
IMO, the `GDateTime` object would be better to be ref at `SircMessage` side, rather than message context side.

File: src/core/app_irc_event.c [hunk 1]
```diff
@@ -43,88 +43,88 @@
 
 static gboolean do_period_ping(gpointer user_data);
 static void add_numeric_error_message(SrnChat *chat, int event, const char
-        *origin, const char **params, int count, const SircMessageContext *context);
+        *origin, const char **params, int count, SircMessageContext *context);
 static void rejoin_all_channels(SrnServer *srv);
 static gboolean rejoin_all_channels_cb(gpointer user_data);
 
 static void irc_event_connect(SircSession *sirc, const char *event,
-        const SircMessageContext *context);
```
Comment (SilverRainZ) [inline src/core/app_irc_event.c:51]: Why we need to remove the const modifier?

**Comment by progval** [comment]
It's not possible to use `g_date_time_ref` on a const object, so `sirc_message_context_get_time` needs to return a non-const `GDateTime`, so it needs to be passed a non-const `SircMessageContext`, which propagates through most of the code. So I just removed the `const` qualifier everywhere.

File: src/core/app_irc_event.c (hunk 1 above)
Comment (progval) [reply src/core/app_irc_event.c:51]: It's not possible to use `g_date_time_ref` on a const object, so `sirc_message_context_get_time` needs to return a non-const `GDateTime`, so it needs to be passed a non-const `SircMessageContext`, which propagates through most of the code. So I just removed the `const` qualifier everywhere.

Comment (SilverRainZ) [reply src/core/app_irc_event.c:51]: I don't think so, `const` only limits you to change the directly content of address (for example, you can't rewrite the `time` field of an `const *SircMessageContext`), but change the content of `time` field is possible. 

See #357.

**Review by SilverRainZ — COMMENTED** [review]
Close, replaced by #357.

**Pull request closed without merging**